rest/oidc_api_test.go

```diff
@@ -1,139 +1,816 @@
+	// relying party is using the authorization code grant type.
+	grantTypeAuthCode grantType = iota + 1
+
+	// grantTypeRefreshToken tells the token endpoint that the
+	// relying party is using the refresh token grant type.
+	grantTypeRefreshToken
+)
+
+// The mockAuthServer represents a mock OAuth2 server for verifying OpenID Connect client code.
+// It is not intended to be used as an actual OAuth 2 server. It lacks many features that would
+// be required in a classic implementation. See https://tools.ietf.org/html/rfc6749 to know more
+// about the OAuth 2.0 Authorization Framework specification.
+type mockAuthServer struct {
+
+	// URL represents base URL of the OAuth2 server of the
+	// form http://ipaddr:port with no trailing slash.
+	URL string
+
+	// server is an underlying HTTP server listening on a system-chosen
+	// port on the local loopback interface, for use in end-to-end HTTP tests.
+	server *httptest.Server
+
+	// The options represents a set of custom options to be injected on mock auth
+	// server to generate the response on demand.
+	options *options
+}
+
+// options represents a set of settings to be configured on mock auth server
+// to simulate the specific authentication behavior during test execution.
+type options struct {
+
+	// OpenID Connect  Provider Issuer URL
+	issuer string
+
+	// wantError represents a specific error expected by the relying party.
+	wantError wantError
+
+	// grantType refers to the way a relying party gets an access token.
+	grantType grantType
+
+	// tokenResponse represents the token response from SG on successful authentication.
+	tokenResponse OIDCTokenResponse
+}
+
+// The newMockAuthServer returns a new mock OAuth Server but doesn't start it.
+// The caller should call Start when needed, to start it up.
+func newMockAuthServer() *mockAuthServer {
+	server := new(mockAuthServer)
+	server.options = new(options)
+	return server
+}
+
+// Start registers mock handlers and starts the mock OAuth server.
+// The caller should call Shutdown when finished, to shut it down.
+func (s *mockAuthServer) Start() {
+	router := mux.NewRouter()
+	router.HandleFunc("/{provider}"+auth.OIDCDiscoveryConfigPath, s.mockDiscoveryHandler).Methods(http.MethodGet)
+	router.HandleFunc("/{provider}/auth", s.mockAuthHandler).Methods(http.MethodGet, http.MethodPost)
+	router.HandleFunc("/{provider}/token", s.mockTokenHandler).Methods(http.MethodPost)
+	s.server = httptest.NewServer(router)
+	s.URL = s.server.URL
+}
+
+// Shutdown shuts down the underlying httptest server and blocks
+// until all outstanding requests on that server have completed.
+func (s *mockAuthServer) Shutdown() {
+	s.server.Close()
+}
+
+// mockDiscoveryHandler mocks the provider discovery endpoint with a mock response.
+// Makes a JSON document available at the path formed by concatenating the string
+// /.well-known/openid-configuration to the Issuer.
+func (s *mockAuthServer) mockDiscoveryHandler(w http.ResponseWriter, r *http.Request) {
```

Comment:
Now that these are functions on mockAuthServer, it would be preferable to remove the 'mock' prefix on the function names (e.g. this can just be discoveryHandler, authHandler, tokenHandler), to remove redundancy and ambiguity.